File: 백준/Silver/9012. 괄호/괄호.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;


public class Main {

    private static final char LEFT = '(';
    private static final char RIGHT = ')';

    private static int T;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        T = Integer.parseInt(br.readLine());

        while(T-- > 0) {
            String input = br.readLine();
            boolean isVPS = isVPS(input);
            if(isVPS) System.out.println("YES");
            else System.out.println("NO");
        }
    }

    private static boolean isVPS(String input) {

        Stack<Character> stack = new Stack<>();

        for(int i = 0; i < input.length(); i++) {

            char cur = input.charAt(i);

            if( cur == LEFT) {
                stack.push(cur);
                continue;
            }

            if(stack.isEmpty() && cur == RIGHT) {
                return false;
            }

            if(cur == RIGHT) {
                stack.pop();
            }
        }

        return stack.isEmpty();
    }
}

File: 백준/Silver/9012. 괄호/괄호.py

def isVPS(input_str: str) -> bool:
    stack = []

    for cur in input_str:
        if cur == '(':
            stack.append(cur)
            continue
        
        if not stack and cur == ')':
            return False
        
        if cur == ')':
            stack.pop()
    
    return not stack

T = int(input())

while T > 0:
    T -= 1
    input_str = input()
    result = isVPS(input_str)
    if result:
        print("YES")
    else:
        print("NO")


